include/rocksdb/cache_allocator.h

```diff
@@ -16,14 +22,35 @@ class CacheAllocator {
   // Name of the cache allocator, printed in the log
   virtual const char* Name() const = 0;
 
-  // Allocate a block of at least size size
+  // Allocate a block of at least size. Has to be thread-safe.
   virtual void* Allocate(size_t size) = 0;
-  // Deallocate previously allocated block
+
+  // Deallocate previously allocated block. Has to be thread-safe.
   virtual void Deallocate(void* p) = 0;
+
   // Returns the memory size of the block allocated at p. The default
   // implementation that just returns the original allocation_size is fine.
   virtual size_t UsableSize(void* /*p*/, size_t allocation_size) const {
     // default implementation just returns the allocation size
     return allocation_size;
   }
 };
+
+// Factory interface to obtain CacheAllocator instances.
+class CacheAllocatorFactory {
```

Comment:
The name "cache allocator" is very confusing to me. Why not call it something like "block allocator" or something like that (because it allocates blocks)?

Reply:
@igorcanadi introduce it in #4437. I don't think "block allocator" is more appropriate since we use it together with `Cache`, not `BlockCache`.

Reply:
The allocator is used inside BlockFetcher, not cache. The class's function has nothing to do with cache at all. It's just that contained by a cache and users can get it from the cache object. Also, "cache allocator" can be easily misunderstood to be an allocator of caches. Actually it allocates memory. I guess MemoryAllocator is a bad idea as we have another allocator interface inherited by arena. How about MemoryAllocatorWrapper?

Reply:
The fact that we only use it in BlockFetcher doesn't affect the allocator abstraction I guess. Just like the `Cache` didn't named `BlockCache` and we later reuse it as also as table cache. The allocator interface can be used later to allocate other memory. I don't see problem with `CacheAllocator` but I'm fine with `MemoryAllocator`. I don't think the class needs to be a `wrapper` though.

Reply:
`MemoryAllocator` feels great to me.

Reply:
@siying I'll make that change in a separate PR.